conformance/utils/kubernetes/apply.go

```diff
@@ -34,17 +34,75 @@ import (
+// them to the Kubernetes cluster.
+type Applier struct {
+	NamespaceLabels    map[string]string
+	ValidListenerPorts []v1alpha2.PortNumber
+}
+
+// prepareGateway adjusts both listener ports and the gatewayClassName. It
+// returns an index pointing to the next valid listener port.
+func prepareGateway(t *testing.T, uObj *unstructured.Unstructured, gatewayClassName string, validListenerPorts []v1alpha2.PortNumber, portIndex int) int {
+	err := unstructured.SetNestedField(uObj.Object, gatewayClassName, "spec", "gatewayClassName")
+	require.NoErrorf(t, err, "error setting `spec.gatewayClassName` on %s Gateway resource", uObj.GetName())
+
+	if len(validListenerPorts) > 0 {
+		listeners, _, err := unstructured.NestedSlice(uObj.Object, "spec", "listeners")
+		require.NoErrorf(t, err, "error getting `spec.listeners` on %s Gateway resource", uObj.GetName())
+
+		for i, uListener := range listeners {
+			require.Less(t, portIndex, len(validListenerPorts), "not enough unassigned valid ports for `spec.listeners[%d]` on %s Gateway resource", i, uObj.GetName())
+
+			listener, ok := uListener.(map[string]interface{})
+			require.Truef(t, ok, "unexpected type at `spec.listeners[%d]` on %s Gateway resource", i, uObj.GetName())
+
+			nextPort := validListenerPorts[portIndex]
+			err = unstructured.SetNestedField(listener, int64(nextPort), "port")
+			require.NoErrorf(t, err, "error setting `spec.listeners[%d].port` on %s Gateway resource", i, uObj.GetName())
+
+			portIndex++
+			listeners[i] = listener
+		}
+
+		err = unstructured.SetNestedSlice(uObj.Object, listeners, "spec", "listeners")
+		require.NoErrorf(t, err, "error setting `spec.listeners` on %s Gateway resource", uObj.GetName())
+	}
+
+	return portIndex
+}
+
+// prepareNamespace adjusts the Namespace labels.
+func prepareNamespace(t *testing.T, uObj *unstructured.Unstructured, namespaceLabels map[string]string) {
+	labels, _, err := unstructured.NestedStringMap(uObj.Object, "metadata", "labels")
+	require.NoErrorf(t, err, "error getting labels on Namespace %s", uObj.GetName())
+
+	if labels == nil {
+		labels = map[string]string{}
+	}
+
+	for k, v := range namespaceLabels {
+		labels[k] = v
+	}
+
+	err = unstructured.SetNestedStringMap(uObj.Object, labels, "metadata", "labels")
+	require.NoErrorf(t, err, "error setting labels on Namespace %s", uObj.GetName())
+}
+
 // MustApplyWithCleanup creates or updates Kubernetes resources defined with the
 // provided YAML file and registers a cleanup function for resources it created.
 // Note that this does not remove resources that already existed in the cluster.
-func MustApplyWithCleanup(t *testing.T, c client.Client, location string, gcName string, cleanup bool) {
+func (a Applier) MustApplyWithCleanup(t *testing.T, c client.Client, location string, gcName string, cleanup bool) {
 	data, err := getContentsFromPathOrURL(location)
 	require.NoError(t, err)
 
 	decoder := yaml.NewYAMLOrJSONDecoder(data, 4096)
+
+	portIndex := 0
```

Comment:
It would be helpful to add a comment here describing the purpose of this index + why it's here any not per-Gateway. I think some may expect the unique ports to apply per Gateway not per manifest file. So if someone has 2 Gateways defined in a manifest with 2 ports each, they may be surprised that they need to define 4 ports and not just 2. (I could also be misinterpreting the code here).

Reply:
Yeah that's the behavior. I'll add a comment that the idea is to keep listener ports unique.